hera_cal/apply_cal.py

```diff
@@ -268,16 +329,31 @@ def apply_cal(data_infilename, data_outfilename, new_calibration, old_calibratio
             if flag_file is not None:
                 data_flags[bl] = np.logical_or(data_flags[bl], ext_flags[bl])
         if redundant_solution:
-            from .redcal import get_reds
-            all_reds = get_reds(data.antpos, pols=data.pols(), bl_error_tol=bl_error_tol)
             calibrate_redundant_solution(data, data_flags, new_gains, new_flags, all_reds, old_gains=old_gains,
                                          old_flags=old_flags, gain_convention=gain_convention)
         else:
             calibrate_in_place(data, new_gains, data_flags=data_flags, cal_flags=new_flags,
                                old_gains=old_gains, gain_convention=gain_convention)
-
-        io.update_vis(data_infilename, data_outfilename, filetype_in=filetype_in, filetype_out=filetype_out,
-                      data=data, flags=data_flags, add_to_history=add_to_history, clobber=clobber, **kwargs)
+        if not redundant_average:
+            io.update_vis(data_infilename, data_outfilename, filetype_in=filetype_in, filetype_out=filetype_out,
+                          data=data, flags=data_flags, add_to_history=add_to_history, clobber=clobber, **kwargs)
+        else:
+            all_red_antpairs = [[bl[:2] for bl in grp] for grp in all_reds if grp[-1][-1] == hd.pols[0]]
+            hd.update(data=data, flags=data_flags, **kwargs)
+            # by default, weight by nsamples (but not flags). This prevents spectral structure from being introduced
+            # and also allows us to compute redundant averaged vis in flagged channels (in case flags are spurious).
+            if no_red_weights:
+                redundant_weights = copy.deepcopy(data_nsamples)
+                for bl in data_flags:
+                    if np.all(data_flags[bl]):
+                        redundant_weights[bl][:] = 0.
+
+            utils.red_average(hd, reds=all_red_antpairs, inplace=True, wgts=redundant_weights)
+            if filetype_out == 'uvh5':
+                # overwrite original outfile with
+                hd.write_uvh5(data_outfilename, clobber=clobber)
```

Comment:
I think you should pass through ` **kwargs`

Reply:
added